src/gc-heap-snapshot.cpp

```diff
@@ -465,9 +472,104 @@ void _record_gc_just_edge(const char *edge_type, Node &from_node, size_t to_idx,
         to_idx // to
     });
 
+    g_snapshot->node_parents.insert(make_pair(to_idx, g_snapshot->node_ptr_to_index_map[(void*)from_node.id]));
+
     g_snapshot->num_edges += 1;
 }
 
+void downsample_heap_snapshot(HeapSnapshot &snapshot, double sample_rate) {
+    // Downsample the heap snapshot by randomly sampling 1/1000th of the nodes, and then
+    // keeping all the edges and nodes from those nodes up to the roots. This is a
+    // simple way to get a representative sample of the heap.
+
+    // We ignore whether or not a node is reachable from the root. Even if it isn't, we can
+    // still take the node, and we'll just keep all nodes going up from it until we reach
+    // a dead end.
+
+    std::cout << "Downsampling heap snapshot, sample rate: " << sample_rate << "\n";
+    std::cout << snapshot.nodes.size() << " original nodes\n";
+
+    if (sample_rate >= 1.0) {
+        std::cout << "Nothing to do." << "\n";
+        return;
+    }
+
+    //HeapSnapshot &snapshot
+
+    // First, sample nodes
+    vector<size_t> sampled_node_idxs;
+    size_t num_nodes = snapshot.nodes.size();
+    for (size_t i = 0; i < num_nodes; i++) {
+        auto r = (static_cast<double>(rand()) / static_cast<double>(RAND_MAX));
+        if (r <= sample_rate) {
+            sampled_node_idxs.push_back(i);
+        }
+    }
+
+    std::cout << sampled_node_idxs.size() << " sampled nodes\n";
+
+    if (sampled_node_idxs.size() >= 100000) {
+        std::cout << "Skipping sampling, it would be too slow. Pick a smaller sample size!\n";
+        return;
+    }
+
+    // Then, find all the parent nodes and edges reachable from those sampled nodes
+    DenseMap<size_t, Node> node_id_to_node_map;
+    for (auto node_idx : sampled_node_idxs) {
+        int depth = 0;
+        while (true) {
+            depth += 1;
+            auto &node = snapshot.nodes[node_idx];
+            auto parent_node_pair = snapshot.node_parents.find(node_idx);
+            if (parent_node_pair == snapshot.node_parents.end()) {
+                break;
+            }
+            auto parent_idx = parent_node_pair->second;
+            auto &parent = snapshot.nodes[parent_idx];
+            node_id_to_node_map.insert(make_pair(node.id, node));
+            auto &new_node = node_id_to_node_map[node.id];
+            node_id_to_node_map.insert(make_pair(parent.id, parent));
+            auto &new_parent = node_id_to_node_map[parent.id];
+            for (auto &edge : node.edges) {
+                if (edge.to_node == node_idx) {
+                    new_node.edges.push_back(edge);
+                }
```

Comment:
I don't really know why we are doing this

Reply:
I'm trying to rebuild the nodes from the root to the child, with edges pointing from each node along that path